forge/lang/expander.rkt

```diff
@@ -724,6 +751,41 @@
        (syntax/loc stx (begin block.test-decls ...))
        (syntax/loc stx (begin)))]))
 
+
+; PropertyWhereDecl : PROPERTY-TOK Name OF-TOK Name Expr WHERE-TOK TEST-CONSTRUCT*
+(define-syntax (PropertyWhereDecl stx)
+  (syntax-parse stx
+  [((~literal PropertyWhereDecl) 
+                              (~and constraint_type (~or "overconstraint" "underconstraint"))  
+                              prop_name:NameClass
+                              "of"
+                              pred_name:NameClass
+                              prop_expr:BlockClass 
+                              (~optional scope:ScopeClass)
+                              (~optional bounds:BoundsClass)
+                               "where"
+                              where-blocks:TestConstructClass ...  ) 
+
+  (with-syntax ([test_name (make-temporary-name stx)]
+                [constraint_type (datum->syntax #'constraint_type
+                                          (string->symbol (syntax->datum #'constraint_type)))]
+                ;;; Overconstraint : Prop => Pred
+                ;;; Underconstraint Pred => Prop
+
+                [imp_total (syntax-e (if (equal? (syntax-e #'constraint_type) "overconstraint") 
```

Comment:
Whoa, there are 2 very confusing things here:

1. `(syntax-e ....)` on the outside --- nvm for now but we'll fix before merging
2. The `if` is comparing a symbol to a string.

Is there a test for this `if` hitting the true case?

Reply:
Yeah, it was surprising to me too but this is tested by the fact that overconstraints and underconstraints test different things (tests/property-where)

Reply:
Ah --- the definition of `constraint_type` in this `with-syntax` is not getting used. Delete it.

`with-syntax*` would've used it

Reply:
Another possibility is to use `#:with` in `syntax-parse`'s clause.  I consider it to be better than both `with-syntax` and `with-syntax*` stylistically, since it doesn't introduce further indentation.

Reply:
Yes `#:with` is much better